### PR TITLE
notify mod on new user articles

### LINKED ISSUE #19295: Round Robin Tweak to include new user published articles
## Is your feature request related to a problem? Please describe.

As an Admin, I’d like to make use of the “Round Robin” notifications to have “Trusted” members receive notifications when new users are making their first posts.

This will be a layer to both ensure that spam and low quality posts have more eyeballs in addition to our other precautions in place, but also that GOOD content from new users is most able to be evaluated and help gain traction on the site — helping ensure overall quality of content on the site, especially for authors who need traction.

## Describe the solution you'd like

We’ve had a [long-standing #TODO](https://github.com/forem/forem/blob/main/app/models/notification.rb#L104) to update our moderator notifications settings to include notifications for published posts by new members.

The logic behind our comment notification (at least partly) [lives here](https://github.com/forem/forem/blob/main/app/models/comment.rb#L211).  We want to provide the same logic for Articles, where it sends out a notification if it’s either the 1st or 2nd published post by that user.

## Additional context

While `user.comments_count` is sufficient for fulfilling the comments logic, we may be better off using the currently available `nth_published_by_author` functionality on `articles` because `user.articles_count` does not differentiate between published and draft. This is just an implementation detail, but can be a heads up as an idea on functionality available to speed this up.

This functionality was most recently addressed in a tweak to frequency and targeting:

* [https://github.com/forem/forem/pull/18995](https://github.com/forem/forem/pull/18995)
* [https://github.com/forem/forem/pull/19026](https://github.com/forem/forem/pull/19026)

### PR DESCRIPTION
**Pull request by dukegreene**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

Adding Articles to the moderation notification worker,  and sending the notification for the first couple articles per author. This change takes the current approach to notifying moderators of new comments by a user, and expands it to notify on article publication.

Added more tests to cover more of the new code. Refactoring suggestions welcomed and appreciated! Not sure how Rails-y the polymorphic behavior of mod-"notifiable" Comments/ Articles should be, attempted to extend the existing comment notification behavior in a reasonable way and allowed for some things to be repeated rather than abstracted (eg the mod_comment_user and mod_article_user methods could take a `type` string argument and fold together into mod_notifiable_user, but I'm not certain they `should` in the name of keeping helper methods clear and readable in-place.

## Related Tickets & Documents

- Related Issue #19295
- Closes #19295

## QA Instructions, Screenshots, Recordings

In my local environment, my first user is a trusted mod. Mod notifications only go out if the user has reacted to a post recently, so I started by clicking a heart as a mod.

Then, in `rails console`, I used `User.first.update(last_moderation_notification: "2023-05-11 17:41:59.260460")` between notification tests to make sure that mod would still be searchable according to the query we use.

Then I tested mod notification by 
  - opening a different browser (eg my mod was using chrome and my users were in Safari), 
  - logging in as a first-time user, and 
  - posting an article or two as that user.
  - Returning to the mod's browser, I clicked the bell icon top right to navigate to /notifications.

Expected behavior is that the first two posts by a user ping the mod with the welcome message (mentioning the post is from a new user), posts after the second published post are ignored.

Drafting a post should not notify the mod.

## Added/updated tests?

- [X ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] What gif best describes this PR or how it makes you feel?

the rock saying "finally"

### CODE REVIEW

File: app/services/notifications/moderation/send.rb [hunk 1]
```diff
@@ -14,7 +14,7 @@ def initialize(moderator, notifiable)
         @notifiable = notifiable
       end
 
-      delegate :user_data, :comment_data, to: Notifications
+      delegate :user_data, :comment_data, :article_data, to: Notifications
 
       def call
         # notifiable is currently only comment
```
Comment (Ridhwana) [inline app/services/notifications/moderation/send.rb:20]: Should we remove this comment now that notifiable is Article and Comment?

File: app/views/notifications/_article.html.erb [hunk 2]
```diff
@@ -3,13 +3,29 @@
...
+    <% if notification.action == "Published" %>
+      <header class="mb-4">
+        <h2 class="fs-base fw-normal">
+          <%= message_user_acted_maybe_org(json_data, "views.notifications.new_post.verb_html", if_org: "views.notifications.new_post.if_org_html") %>
+        </h2>
+        <p class="lh-tight"><small class="fs-s color-base-60"><%= time_ago_in_words json_data["article"]["published_at"], scope: :"datetime.distance_in_words_ago" %></small></p>
+      </header>
+
+      <%= render "notifications/shared/article_preview", json_data: json_data, notification: notification, context: "default" %>
+    <% elsif notification.action == "Moderation" %>
```
Comment (Ridhwana) [inline app/views/notifications/_article.html.erb:15]: (question): Will the `notification.action` be set to `"Moderation"` for only for the welcome email?

**Review by Ridhwana — APPROVED** [review]
Whoop this looks good 🎉

Screenshot 2023-05-18 at 10 57 57

I left a few suggestion/nitpicks/comments but I don't think any are blocking.

File: app/models/notification.rb [hunk 3]
```diff
@@ -101,11 +101,13 @@ def send_welcome_notification(receiver_id, broadcast_id)
...
 
     def send_moderation_notification(notifiable)
-      # TODO: make this work for articles in the future. only works for comments right now
-      return unless notifiable.commentable
-      return if UserBlock.blocking?(notifiable.commentable.user_id, notifiable.user_id)
+      if notifiable.instance_of?(Comment)
+        return if UserBlock.blocking?(notifiable.commentable.user_id, notifiable.user_id)
+      else
+        return unless [Comment, Article].include?(notifiable.class)
+      end
```
[the remark this answers is not in the document]
Comment (dukegreene) [reply app/models/notification.rb:108]: Yeah I think that's les tangly, thanks!

File: app/views/notifications/_article.html.erb (hunk 2 above)
Comment (dukegreene) [reply app/views/notifications/_article.html.erb:15]: Yes, currently the only moderator notification on an article is when it's the user's first or second published article. The method that does that check is new to this PR, and it's the only one attaching the "Moderation" value in the notification's action key, so only the welcome should be hitting this part of the view.

**Comment by mirie** [comment]
@dukegreene I am unsure why the Uffizzi deploy is not working -- do you have a sense? Also @lboogie04 or @Ridhwana not sure if you QA-ed locally so wanted to ask so I could understand that as well! Thanks!

**Comment by Ridhwana** [comment]
> @dukegreene I am unsure why the Uffizzi deploy is not working -- do you have a sense? Also @lboogie04 or @Ridhwana not sure if you QA-ed locally so wanted to ask so I could understand that as well! Thanks!

I QA'ed locally as it was much easier to set up users as per QA instructions where they meet a specific criteria to be assigned as the notified mod. Also from my understanding this feature would be difficult to test on Uffizi as i didn't think you could invite/create new users on Uffizi which is a key part of testing this PR.

**Comment by mirie** [comment]
Yeah, it's a little more manual in terms of adding users! Okay @dukegreene whenever you're ready -- looks like good to merge

**Pull request merged**